App/Services/CredentialManager.cs

```diff
@@ -0,0 +1,247 @@
+        {
+            throw new InvalidOperationException("Could not connect to or verify Coder server", e);
+        }
+
+        WriteCredentials(new RawCredentials
+        {
+            CoderUrl = coderUrl,
+            ApiToken = apiToken,
+        });
+
+        UpdateState(new CredentialModel
+        {
+            State = CredentialState.Valid,
+            CoderUrl = coderUrl,
+            ApiToken = apiToken,
+        });
+    }
+
+    public void ClearCredentials()
+    {
+        NativeApi.DeleteCredentials(CredentialsTargetName);
+        UpdateState(new CredentialModel
+        {
+            State = CredentialState.Invalid,
+            CoderUrl = null,
+            ApiToken = null,
+        });
+    }
+
+    private void UpdateState(CredentialModel newModel)
+    {
+        _latestCredentials = newModel;
+        CredentialsChanged?.Invoke(this, _latestCredentials.Clone());
+    }
+
+    private RawCredentials? ReadCredentials()
+    {
+        var raw = NativeApi.ReadCredentials(CredentialsTargetName);
+        if (raw == null) return null;
+
+        RawCredentials? credentials;
+        try
+        {
+            credentials = JsonSerializer.Deserialize<RawCredentials>(raw);
+        }
+        catch (JsonException)
+        {
+            return null;
+        }
+
+        if (credentials is null || string.IsNullOrWhiteSpace(credentials.CoderUrl) ||
+            string.IsNullOrWhiteSpace(credentials.ApiToken)) return null;
+
+        return credentials;
+    }
+
+    private void WriteCredentials(RawCredentials credentials)
+    {
+        var raw = JsonSerializer.Serialize(credentials);
+        NativeApi.WriteCredentials(CredentialsTargetName, raw);
+    }
+
+    private class RawCredentials
+    {
+        public required string CoderUrl { get; set; }
+        public required string ApiToken { get; set; }
+    }
+
+    private static class NativeApi
+    {
+        private const int CredentialTypeGeneric = 1;
+        private const int PersistenceTypeLocalComputer = 2;
+        private const int ErrorNotFound = 1168;
+
+        public static string? ReadCredentials(string targetName)
+        {
+            if (!CredReadW(targetName, CredentialTypeGeneric, 0, out var credentialPtr))
+            {
+                var error = Marshal.GetLastWin32Error();
+                if (error == ErrorNotFound) return null;
+                throw new InvalidOperationException($"Failed to read credentials (Error {error})");
+            }
+
+            try
+            {
+                var cred = Marshal.PtrToStructure<CREDENTIAL>(credentialPtr);
+                return Marshal.PtrToStringUni(cred.CredentialBlob, cred.CredentialBlobSize / sizeof(char));
+            }
+            finally
+            {
+                CredFree(credentialPtr);
+            }
+        }
+
+        public static void WriteCredentials(string targetName, string secret)
+        {
+            if (Encoding.Unicode.GetByteCount(secret) > 512)
```

Comment:
why is 512 our maximum size?

Reply:
512 is the maximum size of the credentials stored using this API: https://learn.microsoft.com/en-us/windows/win32/api/wincred/ns-wincred-credentiala

Reply:
Fair enough, but...

> The size, in bytes, of the CredentialBlob member. This member cannot be larger than CRED_MAX_CREDENTIAL_BLOB_SIZE (5*512) bytes.

a) shouldn't it be 5*512 = 2560
b) is there any way to grab that constant and use it directly? Seems like it would be more clear

Reply:
Ok, hardcode, but would be nice to make a local `const` with the same name for clarity